# Design of YASIM

## Definitions and Terminologies

Here lists definitions and terminologies used in API docs, etc.

Low-Level Read Generators (LLRG)
: A program that simulates the behavior of real sequencers. It can be abstracted to a step, which takes input sequence, sequencing depth, and some error profiles, and generates simulated sequencing data with its satistical parameters (e.g., read length, truncation, GC bias, etc.) like data generated by real sequencers.

LLRG Python Adapter
: A Python class that wraps LLRG, or LLRG Shell Adapter, to integrate it into the main simulation process. It is called by LLRG UI (introduced below).

LLRG Shell Adapter
: A Shell script that handles installation and maintenance of LLRGs.

LLRG User Interface (LLRG UI)
: A Command Line Interface (CLI) provided by YASIM, allowing users to produce simulated reads using their preferred LLRGs.

## Steps of Simulation

### Input: reference transcriptome (Human or _C. Elegans_, given by a GTF file

Reference GTF can be downloaded at UCSC, NCBI, Ensembl or WormBase (for _C. Elegans_). Support of GFF3 format is purposed **but not implemented**.

| Name    | Genome FASTA                                                 | Genome GTF                                                   | cDNA FASTA                                                   |
| ------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| UCSC    | [hg38.p13](https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz) | [ncbiRefSeq](https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/genes/hg38.ncbiRefSeq.gtf.gz) | NA                                                           |
| Ensembl | [GRCh38.105](http://ftp.ensembl.org/pub/release-105/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz) | [GRCh38.105](http://ftp.ensembl.org/pub/release-105/gtf/homo_sapiens/Homo_sapiens.GRCh38.105.gtf.gz) | [GRCh38.105](http://ftp.ensembl.org/pub/release-105/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz) |
| NCBI    | [GRCh38.p14](https://ftp.ncbi.nlm.nih.gov/genomes/genbank/vertebrate_mammalian/Homo_sapiens/reference/GCA_000001405.29_GRCh38.p14/GCA_000001405.29_GRCh38.p14_genomic.fna.gz) | [GRCh38.p14](https://ftp.ncbi.nlm.nih.gov/genomes/genbank/vertebrate_mammalian/Homo_sapiens/reference/GCA_000001405.29_GRCh38.p14/GCA_000001405.29_GRCh38.p14_genomic.gtf.gz) | NA                                                           |

The reference gene GTF will be parsed into a Gene-Transcript-Exon Three-Tier structure and cached into a `.gvpkl.xz` file.

### `alternativr_splicing` \& `dge:` Simulate AS events and DGE based on a statistical model given by mining raw data and works of literature

%TODO: This part is waiting for satistical tests.

```{note}
The reason why we use "coverage" instead of more formal parameters like RPKM or TPM is that most LLRGs (introduced below) are designed for DNA-Seq simulation and accept coverage only (a minority also accepts a number of reads).
```

The simulated DGE will be written into a `depth.tsv` file with `transcript_id` and `coverage`, which is required for the `transcribe` step introduced below.

```{warning}
The generated coverage is **NOT** number of reads generated! It cannot be used as ground truth to assess quantification software!
```

The ground truth GTF, which should be a Gene-Transcript-Exon Three-Tier structure and cached into a `.gvpkl.xz` file, will be transcribed into cDNA. The ground truth GTF will be in Ensembl format and can be parsed by GTF utilities like GFFCompare.

### `transcribe`: Transcribe GTF into ground truth cDNA

A FASTA of all cDNA and a directory with all cDNAs in separate FASTA will be produced. The cDNA FASTAs will be stored in a special directory structure to save depth information without relying on `depth.tsv` file generated in DGE simulation.

The transcribe function can also be used to generate reference cDNAs for software like Salmon.

```{note}
There are differences between transcribed cDNA and Ensembl-provided cDNA. Ensembl-provided cDNA does not include small features like lncRNA, while `yasim` transcribed cDNA includes all transcripts and exons inside GTF.

% TODO: Correct GTF for cDNA data.
```

### Use Low-Level Read-Generators (LLRGs) like [pbsim2](https://github.com/yukiteruono/pbsim2) or [dwgsim](https://github.com/nh13/dwgsim) to generate raw reads for Next- or Third-Generation Sequencing

This step calls LLRG UIs to "sequence" simulated cDNAs to simulated reads with depth. A ground truth including the actual number of reads generated will be produced.
