# Upstream Simulators

- varsim
  - Python
  - SNP, Indel, SV
  - dwgsim, art
  - <https://github.com/bioinform/varsim.git>
  - Mu, J. C., Mohiyuddin, M., Li, J., Bani Asadi, N., Gerstein, M. B., Abyzov, A., Wong, W. H., & Lam, H. Y. (2015). VarSim: a high-fidelity simulation and validation framework for high-throughput genome sequencing with cancer applications. Bioinformatics (Oxford, England), 31(9), 1469–1471. <https://doi.org/10.1093/bioinformatics/btu828>
- sinc
  - C
  - SNP, Indel, CNV
  - sinc
  - <https://sourceforge.net/projects/sincsimulator/>
  - Pattnaik, S., Gupta, S., Rao, A. A., & Panda, B. (2014). SInC: an accurate and fast error-model based simulator for SNPs, Indels and CNVs coupled with a read generator for short-read sequence data. BMC bioinformatics, 15, 40. <https://doi.org/10.1186/1471-2105-15-40>
- simuG
  - Perl
  - SNP, Indel, CNV, inversion, translocations
  - simuG
  - <https://github.com/yjx1217/simuG>
  - Yue, J. X., & Liti, G. (2019). simuG: a general-purpose genome simulator. Bioinformatics (Oxford, England), 35(21), 4442–4444. <https://doi.org/10.1093/bioinformatics/btz424>
- XomeBlender
  - Python, R
  - ?
  - SNP, Indel, CNV
  - <https://github.com/rsemeraro/XomeBlender>
  - Semeraro, R., Orlandini, V., & Magi, A. (2018). Xome-Blender: A novel cancer genome simulator. PloS one, 13(4), e0194472. <https://doi.org/10.1371/journal.pone.0194472>

