---
orphan: true
---

# Project Proposal

2022-03-04, Zhejian YU, Yaqi SU, Ruihong YUAN

Alternative splicing (AS) is a molecular mechanism that modifies pre-mRNA constructs prior to translation. This process can produce a diversity of mRNAs from a single gene by arranging coding sequences (exons) from recently spliced RNA transcripts into different combinations. The mechanisms of AS help to explain how one gene can be encoded into numerous proteins with various functions. This complexity helps drive the cellular differentiation and diversity observed throughout biology.

RNA-Seq is commonly used for AS identification and numerous tools are available for this kind of analysis. To objectively assess their sensitivity and specificity together with scalability, gold standard datasets are required. Since the true number and types of AS events (such as, exon skipping, intron retention etc) in an experimental dataset is not known, suitable datasets need to be simulated. Moreover, Third-Generation Sequencing (TGS), which is developed in recent years, allows sequencing of single transcript without splitting to short reads. With this method, the transcribed exons can be better preserved, allowing us to perform AS analysis more accurately.

At present, there is a lack of appropriate approach (software) to perform the simulation of AS events in TGS data, and here we want to develop such a simulation method to better aid the further analysis of AS events.

## Proposed Workflows

The simulation process mainly consists of the following steps:

- Input: reference transcriptome (Human or _C. Elegans_, given by a GTF file).
- Simulate AS events based on a statistical model given by mining raw data and literatures.
- Output AS events as ground truth GTF, and transcribe GTF into ground truth cDNA.
- Use Low-Level Read-Generators (LLRGs) like [pbsim2](https://github.com/yukiteruono/pbsim2) or [dwgsim](https://github.com/nh13/dwgsim) to generate raw reads forNext- or Third-Generation Sequencing.
- Align to reference genome using a spliced aligner and call alternative splicing events using software like [Stringtie](https://github.com/gpertea/stringtie) or [Freedie](http://github.com/vpc-ccg/freddie).
- Compare the results with ground truth with [gffcompare](https://github.com/gpertea/gffcompare).

## Review of Existing Alternative Splicing Simulators

- [ASimulatoR](https://github.com/biomedbigdata/ASimulatoR) is a NGS RNA-Seq AS Generator, with builtin LLRG.
  - Problems:
    - This simulator can only simulate NGS reads.
    - This simulator uses built-in LLRG, making it not expandable to new generations of sequencers.
  - Highlights:
    - Provides a statistical model for simulating differentially expressed genes (DEG) and AS events.
- [RNASeqEval](https://github.com/kkrizanovic/RNAseqEval) is a general-purpose RNA-Seq AS Generator.
  - Problems:
    - It uses Python 2 syntax and some ill-written Python libraries.
    - Its model for generating AS events and DEG is too simple to model realistic biological events.
  - Highlights:
    - Ground truth is gffcompare-readable.

## What Our New Simulator Will Bring About

- Enable the simulation of AS events in TGS data.
- The new simulator will be compatible with upstream third-party LLRGs. You may add your favorite LLRG like [ART](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3278762/) for Rosche 454 machines.
- The new simulator will be more portable. It may be installed with any computer that have a Python interpreter, with minimal dependence on other Python packages.
